File: tests/libjournal/channel.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <p2/journal.h>
// support
#include <cassert>


// type aliases
template <bool stateV = true>
using inventory_t = pyre::journal::inventory_t<stateV>;

template <typename severityT, typename inventoryT>
using channel_t = pyre::journal::channel_t<severityT, inventoryT>;


// severity stub
class severity_t :
    public channel_t<severity_t, inventory_t<true>>
{
    // types
public:
    using channel_type = channel_t<severity_t, inventory_t<true>>;

    // metamethods
public:
    // index initialization is required...
    severity_t(const name_type &);
};

// stub implementation
severity_t::severity_t(const name_type & name) :
    channel_type(name)
{}


// exercise the channel state interface
int main() {
    // make a channel
    severity_t channel("test.channel");

    // verify its name
    assert (channel.name() == "test.channel");
    // its state
    assert (channel.state() == true);
    // and again using the conversion to bool
    assert (channel);
    // verify that the default state is as expected
    assert (channel.defaultState() == true);

    // deactivate it
    channel.deactivate();
    // and check
    assert (channel.state() == false);

    // all done
    return 0;
}


// end of file

File: tests/journal.pkg/channel.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# channel state interface
def test():
    """
    Exercise the channel state interface
    """
    # get the package
    from j2.Channel import Channel

    # derive a severity
    class Severity(Channel, inventory_type=Channel.disabled_type):
        """
        A sample derivation
        """

    # make one
    channel = Severity(name="test.channel")

    # verify its name
    assert channel.name == "test.channel"
    # its state
    assert channel.state == False
    # and again using the conversion to bool
    assert not channel

    # activate it
    channel.activate()
    # and check
    assert channel.state == True

    # all done
    return


# main
if __name__ == "__main__":
    # prohibit the journal bindings
    journal_no_libjournal = True
    # run the test
    test()


# end of file
